agent/consul/client.go

```diff
@@ -191,6 +191,34 @@ func (c *Client) Shutdown() error {
 	return nil
 }
 
+func (c *Client) ReconnectSerfWithNewNodeName(oldNodeName string) error {
```

Comment:
Don't you need a mutex during this to prevent operations on a shutdown serf instance ?

Reply:
I would think the reload is probably already protected from concurrent access, but not sure, deserves verification, good catch

Reply:
The state synchronization is indeed protected by two mutexes during reload: https://github.com/hashicorp/consul/blob/1e67f6bada50aca01a4948662fd1c7737194615b/agent/agent.go#L3561

Reply:
PauseSync() stop the agent/server synchronization, it does not protect serf access. Note that I'm not sure it's a problem, but it could lead to some weird behaviors/crashes.

Reply:
You're right, I should acquire the shutdownLock has it's done when the node simply shuts down. Doing it now.

Reply:
Done